### ISSUE TITLE
【ERROR】After upgraded the software version, there was a problem with the database

### ISSUE THREAD

**Issue by 9k001** [immich-app/immich#10768]
### The bug

```
QueryFailedError: column "embedding" of relation "smart_search" does not exist
    at PostgresQueryRunner.query (/usr/src/app/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:219:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async AddFaceSearchRelation1718486162779.up (/usr/src/app/dist/migrations/1718486162779-AddFaceSearchRelation.js:17:9)
    at async MigrationExecutor.executePendingMigrations (/usr/src/app/node_modules/typeorm/migration/MigrationExecutor.js:225:17)
    at async DataSource.runMigrations (/usr/src/app/node_modules/typeorm/data-source/DataSource.js:265:35)
    at async DatabaseRepository.runMigrations (/usr/src/app/dist/repositories/database.repository.js:169:9)
    at async /usr/src/app/dist/services/database.service.js:134:17
    at async /usr/src/app/dist/repositories/database.repository.js:177:23 {
  query: 'ALTER TABLE smart_search ALTER COLUMN embedding SET STORAGE EXTERNAL',
  parameters: undefined,
  driverError: error: column "embedding" of relation "smart_search" does not exist
      at /usr/src/app/node_modules/pg/lib/client.js:526:17
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async PostgresQueryRunner.query (/usr/src/app/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:184:25)
      at async AddFaceSearchRelation1718486162779.up (/usr/src/app/dist/migrations/1718486162779-AddFaceSearchRelation.js:17:9)
      at async MigrationExecutor.executePendingMigrations (/usr/src/app/node_modules/typeorm/migration/MigrationExecutor.js:225:17)
      at async DataSource.runMigrations (/usr/src/app/node_modules/typeorm/data-source/DataSource.js:265:35)
      at async DatabaseRepository.runMigrations (/usr/src/app/dist/repositories/database.repository.js:169:9)
      at async /usr/src/app/dist/services/database.service.js:134:17
      at async /usr/src/app/dist/repositories/database.repository.js:177:23 {
    length: 125,
    severity: 'ERROR',
    code: '42703',
    detail: undefined,
    hint: undefined,
    position: undefined,
    internalPosition: undefined,
    internalQuery: undefined,
    where: undefined,
    schema: undefined,
    table: undefined,
    column: undefined,
    dataType: undefined,
    constraint: undefined,
    file: 'tablecmds.c',
    line: '8216',
    routine: 'ATExecSetStorage'
  },
  length: 125,
  severity: 'ERROR',
  code: '42703',
  detail: undefined,
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'tablecmds.c',
  line: '8216',
  routine: 'ATExecSetStorage'
}

Node.js v20.14.0
api worker exited with code 1
```

### The OS that Immich Server is running on

docker

### Version of Immich Server

v1.107.0

### Version of Immich Mobile App

v1.107.0

### Platform with the issue

- [X] Server
- [ ] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
version: '3.5'
services:
  immich-server:
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION}
    container_name: immich-server
    restart: always
    volumes:
      - ${APP_DATA_HOME}/project/immich/immich-server/upload:/usr/src/app/upload
      - /etc/localtime:/etc/localtime:ro
    ports:
      - "{{ SERVER_MAPPING_PORT }}:${SERVER_PORT}"
    env_file:
      - ${WORKSPACE_HOME}/{{ DeployRepo }}/immich/env/{{ MODE_ENV }}/server.env
    command: [ "start.sh", "immich" ]
    environment:
      - TZ=Asia/Shanghai
      - MODE_ENV={{ MODE_ENV }}

  immich-machine_learning:
    image: ghcr.io/immich-app/immich-machine-learning:${IMMICH_VERSION}
    container_name: immich-machine_learning
    restart: always
    volumes:
      - ${APP_DATA_HOME}/project/immich/machine_learning/cache:/cache
    ports:
      - "{{ MACHINE_LEARNING_MAPPING_PORT }}:${MACHINE_LEARNING_PORT}"
    env_file:
      - ${WORKSPACE_HOME}/{{ DeployRepo }}/immich/env/{{ MODE_ENV }}/machine_learning.env
    environment:
      - TZ=Asia/Shanghai
      - MODE_ENV={{ MODE_ENV }}

networks:
  default:
    name: "{{ MODE_ENV }}"
    external: true
```
```


### Your .env content

```Shell
IMMICH_VERSION=release
SERVER_PORT=3001
MACHINE_LEARNING_PORT=3003
```
```


### Reproduction steps

```bash
1. Upgrade Version
2. Run
3.
...
```

**Comment by mmomjian** [comment]
Please take a pg_dumpall of your database before trying anything: https://immich.app/docs/administration/backup-and-restore/

**Comment by 9k001** [comment]
View Database Version
```
tensorchord/pgvecto-rs             pg14-v0.2.0   sha256:90724186f0a3517cf6914295b5ab410db9ce23190a2d9d0b9dd6463e3fa298f0   2e2cb40c55b8   5 months ago   676MB
```

**Comment by 9k001** [comment]
> Please take a pg_dumpall of your database before trying anything: https://immich.app/docs/administration/backup-and-restore/
OK I have backed up the database.
It looks like the capacity is only 12M

**Comment by alextran1502** [comment]
Are you running a custom database, not from the stack?

**Comment by andyxpert** [comment]
340-ish MB mine, waiting for updates, similar situation as OP

**Comment by alextran1502** [comment]
@andyxpert Can you provide your docker-compose file?

**Comment by andyxpert** [comment]
version: "3.9"

services:
  immich-server:
    container_name: immich_server
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    deploy:
      resources:
        limits:
          cpus: "0.75"
          memory: 4000M
    volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - /home/andyxpert/usb/Pictures:/Pictures:ro
    env_file:
      - stack.env
    environment:
      # Make sure these values get passed through from the env file
      - IMMICH_SERVER_URL
      - IMMICH_WEB_URL
    ports:
      - 2283:3001
    depends_on:
      - redis
      - database
    restart: always

  immich-machine-learning:
    container_name: immich_machine_learning
    image: ghcr.io/immich-app/immich-machine-learning:${IMMICH_VERSION:-release}
    deploy:
      resources:
        limits:
          cpus: "0.75"
          memory: 4000M
    volumes:
      - model-cache:/cache
    env_file:
      - stack.env

  redis:
    container_name: immich_redis
    image: redis:6.2-alpine@sha256:70a7a5b641117670beae0d80658430853896b5ef269ccf00d1827427e3263fa3
    deploy:
      resources:
        limits:
          cpus: "0.75"
          memory: 4000M
    restart: always

  database:
    container_name: immich_postgres
    image: tensorchord/pgvecto-rs:pg14-v0.2.0@sha256:90724186f0a3517cf6914295b5ab410db9ce23190a2d9d0b9dd6463e3fa298f0
    deploy:
      resources:
        limits:
          cpus: "0.75"
          memory: 4000M
    env_file:
      - stack.env
    environment:
      POSTGRES_PASSWORD: ${DB_PASSWORD}
      POSTGRES_USER: ${DB_USERNAME}
      POSTGRES_DB: ${DB_DATABASE_NAME}
      PG_DATA: /var/lib/postgresql/data
    volumes:
      - pgdata:/var/lib/postgresql/data
    restart: always

volumes:
  pgdata:
  model-cache:

**Comment by andyxpert** [comment]
And the stack trace

Starting microservices worker
[Nest] 7  - 07/02/2024, 8:44:46 PM     LOG [Microservices:EventRepository] Initialized websocket server
[Nest] 16  - 07/02/2024, 8:44:46 PM     LOG [Api:EventRepository] Initialized websocket server
Migration "AddFaceSearchRelation1718486162779" failed, error: internal error: entered unreachable code
QueryFailedError: internal error: entered unreachable code
    at PostgresQueryRunner.query (/usr/src/app/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:219:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async AddFaceSearchRelation1718486162779.up (/usr/src/app/dist/migrations/1718486162779-AddFaceSearchRelation.js:23:9)
    at async MigrationExecutor.executePendingMigrations (/usr/src/app/node_modules/typeorm/migration/MigrationExecutor.js:225:17)
    at async DataSource.runMigrations (/usr/src/app/node_modules/typeorm/data-source/DataSource.js:265:35)
    at async DatabaseRepository.runMigrations (/usr/src/app/dist/repositories/database.repository.js:169:9)
    at async /usr/src/app/dist/services/database.service.js:134:17
    at async /usr/src/app/dist/repositories/database.repository.js:177:23 {
  query: '\n' +
    '            CREATE INDEX face_index ON face_search\n' +
    '            USING hnsw (embedding vector_cosine_ops)\n' +
    '            WITH (ef_construction = 300, m = 16)',
  parameters: undefined,
  driverError: error: internal error: entered unreachable code
      at /usr/src/app/node_modules/pg/lib/client.js:526:17
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async PostgresQueryRunner.query (/usr/src/app/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:184:25)
      at async AddFaceSearchRelation1718486162779.up (/usr/src/app/dist/migrations/1718486162779-AddFaceSearchRelation.js:23:9)
      at async MigrationExecutor.executePendingMigrations (/usr/src/app/node_modules/typeorm/migration/MigrationExecutor.js:225:17)
      at async DataSource.runMigrations (/usr/src/app/node_modules/typeorm/data-source/DataSource.js:265:35)
      at async DatabaseRepository.runMigrations (/usr/src/app/dist/repositories/database.repository.js:169:9)
      at async /usr/src/app/dist/services/database.service.js:134:17
      at async /usr/src/app/dist/repositories/database.repository.js:177:23 {
    length: 82,
    severity: 'ERROR',
    code: 'XX000',
    detail: undefined,
    hint: undefined,
    position: undefined,
    internalPosition: undefined,
    internalQuery: undefined,
    where: undefined,
    schema: undefined,
    table: undefined,
    column: undefined,
    dataType: undefined,
    constraint: undefined,
    file: 'utils.rs',
    line: '21',
    routine: undefined
  },
  length: 82,
  severity: 'ERROR',
  code: 'XX000',
  detail: undefined,
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'utils.rs',
  line: '21',
  routine: undefined
}
microservices worker exited with code 1
Detected CPU Cores: 1
Starting api worker
Starting microservices worker

**Comment by bo0tzz** [comment]
We're seeing a few cases of this error, also in #10767.

**Comment by 9k001** [comment]
> Are you running a custom database, not from the stack?



> Are you running a custom database, not from the stack?

My database is deployed independently by docker.
However, the version is consistent with immich's requirements.
```
version: '3.5'
services:
  pgvecto_rs:
    image: tensorchord/pgvecto-rs:pg14-v0.2.0
    container_name: db-pgvecto_rs
    restart: always
    volumes:
      - ${WORKSPACE_HOME}/{{ DeployRepo }}/db/conf/{{ MODE_ENV }}/pgvecto_rs/postgresql.conf:/var/lib/postgresql/data/postgresql.conf
      - ${APP_DATA_HOME}/project/db/pgvecto-rs:/var/lib/postgresql/data
      - ${LOG_HOME}/project/db/pgvecto_rs:/var/lib/postgresql/data/log
    ports:
      - "{{ DB_PGVECTO_RS_MAPPING_PORT }}:${DB_PGVECTO_RS_PORT}"
    env_file:
      - ${WORKSPACE_HOME}/{{ DeployRepo }}/db/env/{{ MODE_ENV }}/pgvecto_rs.env
    environment:
      - TZ=Asia/Shanghai
      - MODE_ENV={{ MODE_ENV }}

networks:
  default:
    name: "{{ MODE_ENV }}"
    external: true
```
Check that the sha version of the database mirror meets immich requirements.

**Comment by 9k001** [comment]
During the upgrade process, I only upgraded the server version. The database has not been updated at all.

**Comment by 9k001** [comment]
> @9k001 did this DB start as a pg vectors image, or was it migrated from vanilla PG?

I confirmed that they are native pg vectors.
Because I know about the last time I upgraded to use the pg vectors vector plug-in.

**Comment by andyxpert** [comment]
Just to isolate the issue, this causes the problem for me, on the db container:
ERROR:  internal error: entered unreachable code
[75] STATEMENT:  
	            CREATE INDEX face_index ON face_search
	            USING hnsw (embedding vector_cosine_ops)
	            WITH (ef_construction = 300, m = 16)
As a workaround, is there a way to temporarily disable face recognition when starting the immich server from env vars ?

**Comment by vitormarcal** [comment]
I have the same issue. I run postgres on host (not a docker container).

```
immich_server              | Detected CPU Cores: 12
immich_server              | Starting api worker
immich_server              | Starting microservices worker
immich_server              | [Nest] 7  - 07/02/2024, 5:17:55 PM     LOG [Microservices:EventRepository] Initialized websocket server
immich_server              | Migration "AddFaceSearchRelation1718486162779" failed, error: column "embedding" of relation "smart_search" does not exist
immich_server              | QueryFailedError: column "embedding" of relation "smart_search" does not exist
immich_server              |     at PostgresQueryRunner.query (/usr/src/app/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:219:19)
immich_server              |     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
immich_server              |     at async AddFaceSearchRelation1718486162779.up (/usr/src/app/dist/migrations/1718486162779-AddFaceSearchRelation.js:17:9)
immich_server              |     at async MigrationExecutor.executePendingMigrations (/usr/src/app/node_modules/typeorm/migration/MigrationExecutor.js:225:17)
immich_server              |     at async DataSource.runMigrations (/usr/src/app/node_modules/typeorm/data-source/DataSource.js:265:35)
immich_server              |     at async DatabaseRepository.runMigrations (/usr/src/app/dist/repositories/database.repository.js:169:9)
immich_server              |     at async /usr/src/app/dist/services/database.service.js:134:17
immich_server              |     at async /usr/src/app/dist/repositories/database.repository.js:177:23 {
immich_server              |   query: 'ALTER TABLE smart_search ALTER COLUMN embedding SET STORAGE EXTERNAL',
immich_server              |   parameters: undefined,
immich_server              |   driverError: error: column "embedding" of relation "smart_search" does not exist
immich_server              |       at /usr/src/app/node_modules/pg/lib/client.js:526:17
immich_server              |       at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
immich_server              |       at async PostgresQueryRunner.query (/usr/src/app/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:184:25)
immich_server              |       at async AddFaceSearchRelation1718486162779.up (/usr/src/app/dist/migrations/1718486162779-AddFaceSearchRelation.js:17:9)
immich_server              |       at async MigrationExecutor.executePendingMigrations (/usr/src/app/node_modules/typeorm/migration/MigrationExecutor.js:225:17)
immich_server              |       at async DataSource.runMigrations (/usr/src/app/node_modules/typeorm/data-source/DataSource.js:265:35)
immich_server              |       at async DatabaseRepository.runMigrations (/usr/src/app/dist/repositories/database.repository.js:169:9)
immich_server              |       at async /usr/src/app/dist/services/database.service.js:134:17
immich_server              |       at async /usr/src/app/dist/repositories/database.repository.js:177:23
```

**Comment by mertalev** [comment]
There are two different issues being discussed here.

If you get the error `column "embedding" of relation "smart_search" does not exist`, see [this](https://github.com/immich-app/immich/discussions/10758#discussioncomment-9942698) comment.

If you get `internal error: entered unreachable code`, I'm unfortunately not sure what causes that issue. You can try bringing down immich, running these commands:
```sql
DROP INDEX face_index;
ALTER TABLE asset_faces ALTER COLUMN embedding SET DATA TYPE real[];
ALTER TABLE asset_faces ALTER COLUMN embedding SET DATA TYPE vector(512);
```
And starting immich back up.

**Comment by dstapp** [comment]
Had exactly the same problem yesterday after updating to v1.107.0 as shown in the stack trace above. Vanilla installation (as in "compose file from the samples"). Currently worked around by downgrading to v1.106.4 and restoring a DB dump from before the update. I assume the two patch versions .1 and .2 don't fix anything in that regard, right? At least I found nothing in the changelog.

**Comment by mio-19** [comment]
> I assume the two patch versions .1 and .2 don't fix anything in that regard, right? At least I found nothing in the changelog.

I think so. I got `unreachable code` error on `.2`.

**Comment by mio-19** [comment]
Is it normal to get 2 errors on the beginning of  restoring a DB dump ?

```
 cat ~/@temp/dump.sql.gz \
| sed "s/SELECT pg_catalog.set_config('search_path', '', false);/SELECT pg_catalog.set_config('search_path', 'public, pg_catalog', true);/g" \
| docker exec -i immich_postgres psql --username=postgres    # Restore Backup
SET
SET
SET
DROP DATABASE
ERROR:  current user cannot be dropped
ERROR:  role "postgres" already exists
ALTER ROLE
SET
SET
SET
SET
SET
```

**Comment by dstapp** [comment]
> Is it normal to get 2 errors on the beginning of restoring a DB dump ?
> 
> ```
>  cat ~/@temp/dump.sql.gz \
> | sed "s/SELECT pg_catalog.set_config('search_path', '', false);/SELECT pg_catalog.set_config('search_path', 'public, pg_catalog', true);/g" \
> | docker exec -i immich_postgres psql --username=postgres    # Restore Backup
> SET
> SET
> SET
> DROP DATABASE
> ERROR:  current user cannot be dropped
> ERROR:  role "postgres" already exists
> ALTER ROLE
> SET
> SET
> SET
> SET
> SET
> ```

According to https://immich.app/docs/administration/backup-and-restore/ you need to drop existing DB data before the restore. Looks to me like that's the problem here.

**Comment by mio-19** [comment]
@dstapp  I have moved old DB_DATA_LOCATION so there is no existing DB data

**Comment by 9k001** [comment]
> If you get the error `column "embedding" of relation "smart_search" does not exist`, see [this](https://github.com/immich-app/immich/discussions/10758#discussioncomment-9942698) comment.

I have restored the use of immich in accordance with the method you gave.

Special attention should be paid to this. According to your subsequent additional comments, the parameters of smart_search have been modified in the original solution. I paste all the commands I execute as follows. Please help me verify them.

```
immich=# SELECT * FROM pg_stat_all_indexes WHERE indexrelname = 'clip_index';
Empty set

immich=# SELECT * FROM pg_stat_all_indexes WHERE indexrelname = 'face_index';
Empty set

immich=# TRUNCATE TABLE smart_search;
TRUNCATE TABLE

immich=# TRUNCATE TABLE asset_faces CASCADE;
NOTICE:  truncate cascades to table "person"
WARNING:  using index "pg_toast_2619_index" despite IgnoreSystemIndexes
WARNING:  using index "pg_toast_2619_index" despite IgnoreSystemIndexes
WARNING:  using index "pg_toast_2619_index" despite IgnoreSystemIndexes
TRUNCATE TABLE

immich=# TRUNCATE TABLE person;
ERROR:  cannot truncate a table referenced in a foreign key constraint
DETAIL:  Table "asset_faces" references "person".
HINT:  Truncate table "asset_faces" at the same time, or use TRUNCATE ... CASCADE.
immich=# ALTER TABLE smart_search
ADD COLUMN IF NOT EXISTS embedding vector(640);
WARNING:  using index "pg_event_trigger_evtname_index" despite IgnoreSystemIndexes
ALTER TABLE

immich=# ALTER TABLE asset_faces
ADD COLUMN IF NOT EXISTS embedding vector(512);
ALTER TABLE

immich=# SET vectors.pgvector_compatibility=on;
SET

immich=# CREATE INDEX IF NOT EXISTS clip_index ON smart_search
    USING hnsw (embedding vector_cosine_ops)
    WITH (ef_construction = 300, m = 16);
CREATE INDEX

immich=# CREATE INDEX IF NOT EXISTS face_index ON asset_faces
    USING hnsw (embedding vector_cosine_ops)
    WITH (ef_construction = 300, m = 16);
CREATE INDEX


```
In addition, are corresponding functions such as intelligent facial recognition automatically triggered? I found the corresponding settings, but I didn't find the execute button.

**Comment by mmomjian** [comment]
> Is it normal to get 2 errors on the beginning of restoring a DB dump ?
> 
> ```
>  cat <PATH HIDDEN> \
> | sed "s/SELECT pg_catalog.set_config('search_path', '', false);/SELECT pg_catalog.set_config('search_path', 'public, pg_catalog', true);/g" \
> | docker exec -i immich_postgres psql --username=postgres    # Restore Backup
> SET
> SET
> SET
> DROP DATABASE
> ERROR:  current user cannot be dropped
> ERROR:  role "postgres" already exists
> ALTER ROLE
> SET
> SET
> SET
> SET
> SET
> ```
> 
> Edit: my backup command was ` docker exec -t immich_postgres pg_dumpall -c -U postgres`

This is normal.

**Comment by bbklopfer** [comment]
`unreachable code` error fixed for me using solution in #10767.

**Comment by andyxpert** [comment]
Right, solution from #10767 fixed my issue. Waited for Alex's confirmation to avoid a DB rebuild.
